File: sdks/python/aether_sdk/workflow/state_machine.py
"""
Workflow State Machine

Provides visual workflow definitions with state transitions
for building long-running processes.

Example:
    >>> from aether_sdk.workflow.state_machine import Workflow, State, Transition, WorkflowExecutor
    >>> workflow = Workflow("approval-flow")
    ...     .state("draft")
    ...     .state("approved", is_final=True)
    ...     .transition("submit", "draft", "approved")
    ...     .build()
    >>> executor = WorkflowExecutor()
    >>> result = await executor.start(workflow, {"request_id": "123"})
"""

from __future__ import annotations

import logging
import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Callable, Dict, Generic, List, Optional, Set, TypeVar

from .types import (
    Duration,
    InvalidTransitionError,
    TransitionHandler,
    TransitionResult,
    WorkflowContext,
    WorkflowError,
    WorkflowResult,
    WorkflowStatus,
    WorkflowSuspendedError,
)

logger = logging.getLogger(__name__)

T = TypeVar("T")


@dataclass
class State:
    """A state in the workflow state machine.

    Attributes:
        name: State name.
        is_initial: Whether this is the starting state.
        is_final: Whether this is a terminal state.
        on_enter: Action executed when entering this state.
        on_exit: Action executed when leaving this state.
        timeout: Optional timeout for this state.
        timeout_transition: Transition to fire on timeout.
        metadata: Arbitrary key-value metadata.
    """

    name: str
    is_initial: bool = False
    is_final: bool = False
    on_enter: Optional[TransitionHandler] = None
    on_exit: Optional[TransitionHandler] = None
    timeout: Optional[Duration] = None
    timeout_transition: Optional[str] = None
    metadata: Dict[str, Any] = field(default_factory=dict)


@dataclass
class Transition:
    """A transition between two states.

    Attributes:
        name: Transition name.
        from_state: Source state name.
        to_state: Target state name.
        guard: Optional predicate that must return ``True`` for the
            transition to proceed.
        action: Optional action executed during the transition.
        metadata: Arbitrary key-value metadata.
    """

    name: str
    from_state: str
    to_state: str
    guard: Optional[Callable[[WorkflowContext], bool]] = None
    action: Optional[TransitionHandler] = None
    metadata: Dict[str, Any] = field(default_factory=dict)


class Workflow(Generic[T]):
    """A workflow definition as a state machine.

    Defines states and transitions between them. Supports method
    chaining for fluent API construction.

    Example:
        >>> wf = Workflow("order-workflow")
        ...     .state("created", is_initial=True)
        ...     .state("shipped", is_final=True)
        ...     .transition("ship", "created", "shipped")
        ...     .on_enter("shipped", notify_shipped)
        ...     .build()

    Type Parameters:
        T: Type of the workflow input.
    """

    def __init__(self, name: str):
        """Initialize the workflow.

        Args:
            name: Workflow name.
        """
        self.name = name
        self._states: Dict[str, State] = {}
        self._transitions: Dict[str, List[Transition]] = {}
        self._initial_state: Optional[str] = None
        self._final_states: Set[str] = set()
        self._current_state: Optional[State] = None
        self._metadata: Dict[str, Any] = {}

    def state(
        self,
        name: str,
        is_initial: bool = False,
        is_final: bool = False,
        timeout: Optional[Duration] = None,
        timeout_transition: Optional[str] = None,
    ) -> Workflow[T]:
        """Add a state to the workflow.

        Args:
            name: State name.
            is_initial: Whether this is the initial state.
            is_final: Whether this is a terminal state.
            timeout: Optional timeout for this state.
            timeout_transition: Transition to take on timeout.

        Returns:
            Self for method chaining.

        Raises:
            WorkflowError: If multiple initial states are defined.
        """
        state = State(
            name=name,
            is_initial=is_initial,
            is_final=is_final,
            timeout=timeout,
            timeout_transition=timeout_transition,
        )
        self._states[name] = state

        if is_initial:
            if self._initial_state is not None:
                raise WorkflowError(
                    f"Multiple initial states: {self._initial_state} and {name}"
                )
            self._initial_state = name

        if is_final:
            self._final_states.add(name)

        self._current_state = state
        return self

    def on_enter(self, state_name: str, handler: TransitionHandler) -> Workflow[T]:
        """Set the on-enter handler for a state.

        Args:
            state_name: Name of an existing state.
            handler: Async or sync callable.

        Returns:
            Self for method chaining.

        Raises:
            WorkflowError: If the state name is not recognized.
        """
        if state_name not in self._states:
            raise WorkflowError(f"Unknown state: {state_name}")
        self._states[state_name].on_enter = handler
        return self

    def on_exit(self, state_name: str, handler: TransitionHandler) -> Workflow[T]:
        """Set the on-exit handler for a state.

        Args:
            state_name: Name of an existing state.
            handler: Async or sync callable.

        Returns:
            Self for method chaining.

        Raises:
            WorkflowError: If the state name is not recognized.
        """
        if state_name not in self._states:
            raise WorkflowError(f"Unknown state: {state_name}")
        self._states[state_name].on_exit = handler
        return self

    def transition(
        self,
        name: str,
        from_state: str,
        to_state: str,
        guard: Optional[Callable[[WorkflowContext], bool]] = None,
    ) -> Workflow[T]:
        """Add a transition between states.

        Args:
            name: Transition name.
            from_state: Source state name.
            to_state: Target state name.
            guard: Optional guard predicate.

        Returns:
            Self for method chaining.

        Raises:
            WorkflowError: If either state is not recognized.
        """
        if from_state not in self._states:
            raise WorkflowError(f"Unknown source state: {from_state}")
        if to_state not in self._states:
            raise WorkflowError(f"Unknown target state: {to_state}")

        transition = Transition(
            name=name,
            from_state=from_state,
            to_state=to_state,
            guard=guard,
        )

        if from_state not in self._transitions:
            self._transitions[from_state] = []
        self._transitions[from_state].append(transition)

        return self

    def with_action(
        self,
        transition_name: str,
        action: TransitionHandler,
    ) -> Workflow[T]:
        """Set the action for a named transition.

        Args:
            transition_name: Name of an existing transition.
            action: Async or sync callable.

        Returns:
            Self for method chaining.

        Raises:
            WorkflowError: If the transition name is not found.
        """
        for transitions in self._transitions.values():
            for t in transitions:
                if t.name == transition_name:
                    t.action = action
                    return self
        raise WorkflowError(f"Unknown transition: {transition_name}")

    def with_metadata(self, key: str, value: Any) -> Workflow[T]:
        """Add metadata to the workflow.

        Args:
            key: Metadata key.
            value: Metadata value.

        Returns:
            Self for method chaining.
        """
        self._metadata[key] = value
        return self

    def build(self) -> Workflow[T]:
        """Validate and return the workflow.

        Returns:
            The validated workflow instance.

        Raises:
            WorkflowError: If no initial state is defined.
        """
        if self._initial_state is None:
            raise WorkflowError("No initial state defined")

        return self

    @property
    def states(self) -> Dict[str, State]:
        """Return a copy of all states.

        Returns:
            A dict mapping state names to :class:`State` objects.
        """
        return dict(self._states)

    @property
    def initial_state(self) -> str:
        """Return the initial state name.

        Returns:
            The name of the initial state.

        Raises:
            WorkflowError: If no initial state is defined.
        """
        if self._initial_state is None:
            raise WorkflowError("No initial state defined")
        return self._initial_state

    def is_final_state(self, state_name: str) -> bool:
        """Check whether a state is a final (terminal) state.

        Args:
            state_name: State name.

        Returns:
            ``True`` if the state is final.
        """
        return state_name in self._final_states

    def get_transitions(self, from_state: str) -> List[Transition]:
        """Return all transitions from a given state.

        Args:
            from_state: Source state name.

        Returns:
            A list of :class:`Transition` objects.
        """
        return self._transitions.get(from_state, [])

    def get_transition(self, from_state: str, name: str) -> Optional[Transition]:
        """Look up a specific transition by name from a state.

        Args:
            from_state: Source state name.
            name: Transition name.

        Returns:
            The matching :class:`Transition`, or ``None``.
        """
        for t in self._transitions.get(from_state, []):
            if t.name == name:
                return t
        return None

    def validate_transition(
        self,
        from_state: str,
        transition_name: str,
        context: WorkflowContext,
    ) -> Optional[Transition]:
        """Validate that a transition is allowed.

        Checks that the transition exists and its guard (if any)
        returns ``True``.

        Args:
            from_state: Current state name.
            transition_name: Name of the transition to validate.
            context: The workflow context (passed to guard predicates).

        Returns:
            The :class:`Transition` if valid, ``None`` otherwise.
        """
        transition = self.get_transition(from_state, transition_name)
        if transition is None:
            return None

        if transition.guard and not transition.guard(context):
            return None

        return transition


class WorkflowExecutor:
    """Executes workflow state machines.

    Manages workflow lifecycle, state transitions, and persistence.

    Example:
        >>> executor = WorkflowExecutor()
        >>> result = await executor.start(workflow, {"order_id": "123"})
        >>> result = await executor.transition(result.workflow_id, "ship")
        >>> status = await executor.get_status(result.workflow_id)
    """

    def __init__(self):
        """Initialize with empty workflow registries."""
        self._workflows: Dict[str, WorkflowContext] = {}
        self._definitions: Dict[str, Workflow] = {}

    async def start(
        self,
        workflow: Workflow[T],
        input: T,
        workflow_id: Optional[str] = None,
    ) -> WorkflowResult:
        """Start a new workflow execution.

        Args:
            workflow: The workflow definition.
            input: Input data for the workflow.
            workflow_id: Optional explicit ID.

        Returns:
            A :class:`WorkflowResult` with the initial status.
        """
        wf_id = workflow_id or str(uuid.uuid4())

        context = WorkflowContext[T](
            workflow_id=wf_id,
            workflow_type=workflow.name,
            current_state=workflow.initial_state,
            input=input,
            started_at=datetime.now(timezone.utc),
            updated_at=datetime.now(timezone.utc),
        )

        self._workflows[wf_id] = context
        self._definitions[wf_id] = workflow

        initial_state = workflow._states.get(workflow.initial_state)
        if initial_state and initial_state.on_enter:
            try:
                await initial_state.on_enter(context)
            except Exception as e:
                logger.error(f"Failed to execute on-enter for initial state: {e}")

        context.add_history_event(
            "workflow_started",
            initial_state=workflow.initial_state,
        )

        return WorkflowResult(
            workflow_id=wf_id,
            status=WorkflowStatus.RUNNING,
            current_state=context.current_state,
            started_at=context.started_at,
        )

    async def transition(
        self,
        workflow_id: str,
        transition_name: str,
        payload: Optional[Dict[str, Any]] = None,
    ) -> TransitionResult:
        """Execute a state transition on a running workflow.

        Args:
            workflow_id: The workflow instance ID.
            transition_name: The transition to execute.
            payload: Optional payload for the transition.

        Returns:
            A :class:`TransitionResult` with the outcome.

        Raises:
            WorkflowError: If the workflow or definition is not found.
            WorkflowSuspendedError: If the workflow is suspended.
            InvalidTransitionError: If the transition is not valid.
        """
        context = self._workflows.get(workflow_id)
        if context is None:
            raise WorkflowError(f"Unknown workflow: {workflow_id}")

        workflow = self._definitions.get(workflow_id)
        if workflow is None:
            raise WorkflowError(f"Unknown workflow definition: {workflow_id}")

        if context.status == WorkflowStatus.SUSPENDED:
            raise WorkflowSuspendedError(workflow_id)

        from_state = context.current_state

        transition = workflow.validate_transition(from_state, transition_name, context)
        if transition is None:
            raise InvalidTransitionError(from_state, transition_name, workflow_id)

        to_state = transition.to_state

        try:
            current_state_def = workflow._states.get(from_state)
            if current_state_def and current_state_def.on_exit:
                await current_state_def.on_exit(context)

            if transition.action:
                await transition.action(context)

            context.current_state = to_state
            context.updated_at = datetime.now(timezone.utc)

            new_state_def = workflow._states.get(to_state)
            if new_state_def and new_state_def.on_enter:
                await new_state_def.on_enter(context)

            context.add_history_event(
                "transition",
                transition=transition_name,
                from_state=from_state,
                to_state=to_state,
            )

            return TransitionResult(
                success=True,
                from_state=from_state,
                to_state=to_state,
            )

        except Exception as e:
            logger.error(f"Transition failed: {e}", exc_info=True)
            context.add_history_event(
                "transition_failed",
                transition=transition_name,
                from_state=from_state,
                error=str(e),
            )

            return TransitionResult(
                success=False,
                from_state=from_state,
                to_state=to_state,
                error=str(e),
            )

    async def suspend(self, workflow_id: str, reason: str = "") -> None:
        """Suspend a running workflow.

        Args:
            workflow_id: The workflow instance ID.
            reason: Optional reason for suspension.

        Raises:
            WorkflowError: If the workflow is not found.
        """
        context = self._workflows.get(workflow_id)
        if context is None:
            raise WorkflowError(f"Unknown workflow: {workflow_id}")

        context.status = WorkflowStatus.SUSPENDED
        context.updated_at = datetime.now(timezone.utc)
        context.add_history_event("suspended", reason=reason)

    async def resume(self, workflow_id: str) -> None:
        """Resume a suspended workflow.

        Args:
            workflow_id: The workflow instance ID.

        Raises:
            WorkflowError: If the workflow is not found or not suspended.
        """
        context = self._workflows.get(workflow_id)
        if context is None:
            raise WorkflowError(f"Unknown workflow: {workflow_id}")

        if context.status != WorkflowStatus.SUSPENDED:
            raise WorkflowError(f"Workflow {workflow_id} is not suspended")

        context.status = WorkflowStatus.RUNNING
        context.updated_at = datetime.now(timezone.utc)
        context.add_history_event("resumed")

    async def cancel(self, workflow_id: str, reason: str = "") -> None:
        """Cancel a running workflow.

        Args:
            workflow_id: The workflow instance ID.
            reason: Optional reason for cancellation.

        Raises:
            WorkflowError: If the workflow is not found.
        """
        context = self._workflows.get(workflow_id)
        if context is None:
            raise WorkflowError(f"Unknown workflow: {workflow_id}")

        context.status = WorkflowStatus.CANCELLED
        context.updated_at = datetime.now(timezone.utc)
        context.add_history_event("cancelled", reason=reason)

    async def get_status(self, workflow_id: str) -> Optional[WorkflowResult]:
        """Get the current status of a workflow.

        Args:
            workflow_id: The workflow instance ID.

        Returns:
            A :class:`WorkflowResult`, or ``None`` if not found.
        """
        context = self._workflows.get(workflow_id)
        if context is None:
            return None

        workflow = self._definitions.get(workflow_id)
        is_final = workflow.is_final_state(context.current_state) if workflow else False

        status = context.status
        if status == WorkflowStatus.RUNNING and is_final:
            status = WorkflowStatus.COMPLETED

        return WorkflowResult(
            workflow_id=workflow_id,
            status=status,
            current_state=context.current_state,
            output=context.variables,
            history=context.history,
            started_at=context.started_at,
            updated_at=context.updated_at,
        )

    def get_available_transitions(self, workflow_id: str) -> List[str]:
        """Return the names of transitions available from the current state.

        Transitions whose guards return ``False`` are excluded.

        Args:
            workflow_id: The workflow instance ID.

        Returns:
            A list of transition names.
        """
        context = self._workflows.get(workflow_id)
        if context is None:
            return []

        workflow = self._definitions.get(workflow_id)
        if workflow is None:
            return []

        transitions = workflow.get_transitions(context.current_state)
        return [t.name for t in transitions if t.guard is None or t.guard(context)]


def workflow(name: str) -> Workflow:
    """Factory function to create a new workflow.

    Args:
        name: Workflow name.

    Returns:
        A new :class:`Workflow` instance.

    Example:
        >>> wf = workflow("approval-flow")
        ...     .state("draft", is_initial=True)
        ...     .state("approved", is_final=True)
        ...     .transition("submit", "draft", "approved")
        ...     .build()
    """
    return Workflow(name)


__all__ = [
    "State",
    "Transition",
    "Workflow",
    "WorkflowExecutor",
    "workflow",
]

File: sdks/javascript/src/workflow/state_machine.ts
/**
 * Workflow State Machine.
 *
 * Provides state machine definitions with transitions, guards,
 * and lifecycle hooks for building long-running processes.
 *
 * @module aether/workflow/state_machine
 */

import {
  WorkflowStatus,
  WorkflowContext,
  WorkflowResult,
  TransitionResult,
  WorkflowError,
  InvalidTransitionError,
  WorkflowSuspendedError,
  StateMachineEventType,
  TransitionHandler,
} from './types';

/**
 * A state in the workflow state machine.
 */
export class StateDefinition {
  name: string;
  isInitial: boolean = false;
  isFinal: boolean = false;
  onEnter?: TransitionHandler;
  onExit?: TransitionHandler;
  metadata: Record<string, unknown> = {};

  constructor(name: string) {
    this.name = name;
  }
}

/**
 * A transition between two states.
 */
export class StateTransition {
  name: string;
  fromState: string;
  toState: string;
  guard?: (context: WorkflowContext) => boolean;
  action?: TransitionHandler;
  metadata: Record<string, unknown> = {};

  constructor(
    name: string,
    fromState: string,
    toState: string,
    guard?: (context: WorkflowContext) => boolean
  ) {
    this.name = name;
    this.fromState = fromState;
    this.toState = toState;
    this.guard = guard;
  }

  /** Set the transition action. Returns `this` for chaining. */
  withAction(action: TransitionHandler): this {
    this.action = action;
    return this;
  }
}

/**
 * History entry for a state change.
 */
export interface StateHistoryEntry {
  /** Type of event (transition, enter, exit, etc.). */
  type: StateMachineEventType;
  /** State name the event relates to. */
  state: string;
  /** Transition name (for transition events). */
  transition?: string;
  /** Previous state (for transition events). */
  fromState?: string;
  /** Target state (for transition events). */
  toState?: string;
  /** Timestamp of the event. */
  timestamp: Date;
  /** Optional error details. */
  error?: string;
}

/**
 * A workflow definition as a state machine.
 *
 * Defines states and transitions between them. Supports method
 * chaining for fluent API construction.
 *
 * @typeParam T - Type of the workflow input.
 *
 * @example
 * ```typescript
 * const wf = new StateMachineBuilder('order-workflow')
 *   .state('created', { isInitial: true })
 *   .state('shipped', { isFinal: true })
 *   .transition('ship', 'created', 'shipped')
 *   .onEnter('shipped', notifyShipped)
 *   .build();
 * ```
 */
export class StateMachineBuilder<T = unknown> {
  private readonly _states: Map<string, StateDefinition> = new Map();
  private readonly _transitions: Map<string, StateTransition[]> = new Map();
  private _initialState: string | null = null;
  private readonly _finalStates: Set<string> = new Set();
  private readonly _metadata: Record<string, unknown> = {};

  constructor(public readonly name: string) {}

  /**
   * Add a state to the state machine.
   *
   * @param name      - State name.
   * @param options   - State configuration options.
   * @returns `this` for chaining.
   * @throws {WorkflowError} If multiple initial states are defined.
   */
  state(
    name: string,
    options: {
      isInitial?: boolean;
      isFinal?: boolean;
    } = {}
  ): this {
    const def = new StateDefinition(name);
    def.isInitial = options.isInitial ?? false;
    def.isFinal = options.isFinal ?? false;
    this._states.set(name, def);

    if (def.isInitial) {
      if (this._initialState !== null) {
        throw new WorkflowError(
          `Multiple initial states: ${this._initialState} and ${name}`
        );
      }
      this._initialState = name;
    }

    if (def.isFinal) {
      this._finalStates.add(name);
    }

    return this;
  }

  /**
   * Set the on-enter handler for a state.
   *
   * @returns `this` for chaining.
   * @throws {WorkflowError} If the state name is not recognized.
   */
  onEnter(stateName: string, handler: TransitionHandler): this {
    const def = this._states.get(stateName);
    if (def === undefined) {
      throw new WorkflowError(`Unknown state: ${stateName}`);
    }
    def.onEnter = handler;
    return this;
  }

  /**
   * Set the on-exit handler for a state.
   *
   * @returns `this` for chaining.
   * @throws {WorkflowError} If the state name is not recognized.
   */
  onExit(stateName: string, handler: TransitionHandler): this {
    const def = this._states.get(stateName);
    if (def === undefined) {
      throw new WorkflowError(`Unknown state: ${stateName}`);
    }
    def.onExit = handler;
    return this;
  }

  /**
   * Add a transition between states.
   *
   * @param name       - Transition name.
   * @param fromState  - Source state name.
   * @param toState    - Target state name.
   * @param guard      - Optional guard predicate.
   * @returns `this` for chaining.
   * @throws {WorkflowError} If either state is not recognized.
   */
  transition(
    name: string,
    fromState: string,
    toState: string,
    guard?: (context: WorkflowContext) => boolean
  ): this {
    if (!this._states.has(fromState)) {
      throw new WorkflowError(`Unknown source state: ${fromState}`);
    }
    if (!this._states.has(toState)) {
      throw new WorkflowError(`Unknown target state: ${toState}`);
    }

    const t = new StateTransition(name, fromState, toState, guard);
    const existing = this._transitions.get(fromState);
    if (existing) {
      existing.push(t);
    } else {
      this._transitions.set(fromState, [t]);
    }

    return this;
  }

  /**
   * Set the action for a named transition.
   *
   * @returns `this` for chaining.
   * @throws {WorkflowError} If the transition name is not found.
   */
  withAction(transitionName: string, action: TransitionHandler): this {
    for (const transitions of this._transitions.values()) {
      for (const t of transitions) {
        if (t.name === transitionName) {
          t.action = action;
          return this;
        }
      }
    }
    throw new WorkflowError(`Unknown transition: ${transitionName}`);
  }

  /** Add metadata. Returns `this` for chaining. */
  withMetadata(key: string, value: unknown): this {
    this._metadata[key] = value;
    return this;
  }

  /**
   * Validate the state machine definition.
   *
   * @returns `this` for chaining.
   * @throws {WorkflowError} If no initial state is defined.
   */
  build(): this {
    if (this._initialState === null) {
      throw new WorkflowError('No initial state defined');
    }
    return this;
  }

  /** Return a copy of all state definitions. */
  get states(): Map<string, StateDefinition> {
    return new Map(this._states);
  }

  /** Return the initial state name. */
  get initialState(): string {
    if (this._initialState === null) {
      throw new WorkflowError('No initial state defined');
    }
    return this._initialState;
  }

  /** Check whether a state is a final (terminal) state. */
  isFinalState(stateName: string): boolean {
    return this._finalStates.has(stateName);
  }

  /** Return all transitions from a given state. */
  getTransitions(fromState: string): StateTransition[] {
    return this._transitions.get(fromState) ?? [];
  }

  /** Look up a specific transition by name from a state. */
  getTransition(fromState: string, name: string): StateTransition | undefined {
    return this._transitions.get(fromState)?.find(t => t.name === name);
  }

  /**
   * Validate that a transition is allowed.
   *
   * Checks that the transition exists and its guard (if any) returns `true`.
   *
   * @returns The matching {@link StateTransition}, or `undefined` if invalid.
   */
  validateTransition(
    fromState: string,
    transitionName: string,
    context: WorkflowContext
  ): StateTransition | undefined {
    const transition = this.getTransition(fromState, transitionName);
    if (transition === undefined) {
      return undefined;
    }
    if (transition.guard !== undefined && !transition.guard(context)) {
      return undefined;
    }
    return transition;
  }
}

/**
 * Executes workflow state machines.
 *
 * Manages workflow lifecycle, state transitions, and history tracking.
 *
 * @example
 * ```typescript
 * const executor = new StateMachineExecutor();
 * const result = await executor.start(workflow, { orderId: '123' });
 * const tr = await executor.send(result.workflowId, 'ship');
 * const status = await executor.getStatus(result.workflowId);
 * ```
 */
export class StateMachineExecutor {
  private readonly workflows: Map<string, WorkflowContext> = new Map();
  private readonly definitions: Map<string, StateMachineBuilder> = new Map();
  private readonly histories: Map<string, StateHistoryEntry[]> = new Map();

  /**
   * Start a new workflow execution.
   *
   * @typeParam T - Type of the workflow input.
   * @param workflow   - The state machine definition.
   * @param input      - Input data.
   * @param workflowId - Optional explicit ID.
   * @returns A {@link WorkflowResult} with the initial status.
   */
  async start<T>(
    workflow: StateMachineBuilder<T>,
    input: T,
    workflowId?: string
  ): Promise<WorkflowResult> {
    const wfId = workflowId ?? crypto.randomUUID();

    const context = new WorkflowContext<T>(wfId);
    context.workflowType = workflow.name;
    context.currentState = workflow.initialState;
    context.input = input;
    context.startedAt = new Date();
    context.updatedAt = new Date();

    this.workflows.set(wfId, context);
    this.definitions.set(wfId, workflow);
    this.histories.set(wfId, []);

    const initialStateDef = workflow.states.get(workflow.initialState);
    if (initialStateDef?.onEnter) {
      try {
        await Promise.resolve(initialStateDef.onEnter(context));
      } catch (e) {
        // Log but don't fail the start
      }
    }

    this._recordHistory(wfId, {
      type: StateMachineEventType.Enter,
      state: workflow.initialState,
    });

    context.addHistoryEvent('workflow_started', {
      initialState: workflow.initialState,
    });

    return {
      workflowId: wfId,
      status: WorkflowStatus.Running,
      currentState: context.currentState,
      history: [...context.history],
      startedAt: context.startedAt,
    };
  }

  /**
   * Execute a state transition on a running workflow.
   *
   * @param workflowId      - The workflow instance ID.
   * @param transitionName  - The transition to execute.
   * @param payload         - Optional payload for the transition.
   * @returns A {@link TransitionResult} with the outcome.
   * @throws {WorkflowError}           If the workflow is not found.
   * @throws {WorkflowSuspendedError}  If the workflow is suspended.
   * @throws {InvalidTransitionError}  If the transition is not valid.
   */
  async send(
    workflowId: string,
    transitionName: string,
    payload?: Record<string, unknown>
  ): Promise<TransitionResult> {
    const context = this.workflows.get(workflowId);
    if (context === undefined) {
      throw new WorkflowError(`Unknown workflow: ${workflowId}`);
    }

    const workflow = this.definitions.get(workflowId);
    if (workflow === undefined) {
      throw new WorkflowError(`Unknown workflow definition: ${workflowId}`);
    }

    if (context.status === WorkflowStatus.Suspended) {
      throw new WorkflowSuspendedError(workflowId);
    }

    const fromState = context.currentState;

    const transition = workflow.validateTransition(
      fromState,
      transitionName,
      context
    );
    if (transition === undefined) {
      this._recordHistory(workflowId, {
        type: StateMachineEventType.GuardFailed,
        state: fromState,
        transition: transitionName,
        toState: workflow.getTransition(fromState, transitionName)?.toState,
      });
      throw new InvalidTransitionError(fromState, transitionName, workflowId);
    }

    const toState = transition.toState;

    try {
      const fromDef = workflow.states.get(fromState);
      if (fromDef?.onExit) {
        await Promise.resolve(fromDef.onExit(context));
      }

      this._recordHistory(workflowId, {
        type: StateMachineEventType.Exit,
        state: fromState,
      });

      if (transition.action) {
        await Promise.resolve(transition.action(context));
      }

      if (payload) {
        for (const [k, v] of Object.entries(payload)) {
          context.setVariable(k, v);
        }
      }

      context.currentState = toState;
      context.updatedAt = new Date();

      this._recordHistory(workflowId, {
        type: StateMachineEventType.Transition,
        state: toState,
        fromState,
        toState,
        transition: transitionName,
      });

      const toDef = workflow.states.get(toState);
      if (toDef?.onEnter) {
        await Promise.resolve(toDef.onEnter(context));
      }

      this._recordHistory(workflowId, {
        type: StateMachineEventType.Enter,
        state: toState,
      });

      context.addHistoryEvent('transition', {
        transition: transitionName,
        fromState,
        toState,
      });

      return {
        success: true,
        fromState,
        toState,
        timestamp: new Date(),
      };
    } catch (e) {
      const errMsg = e instanceof Error ? e.message : String(e);

      this._recordHistory(workflowId, {
        type: StateMachineEventType.ActionFailed,
        state: fromState,
        transition: transitionName,
        toState,
        error: errMsg,
      });

      context.addHistoryEvent('transition_failed', {
        transition: transitionName,
        fromState,
        error: errMsg,
      });

      return {
        success: false,
        fromState,
        toState,
        error: errMsg,
        timestamp: new Date(),
      };
    }
  }

  /**
   * Check whether a transition is allowed from the current state.
   */
  canTransition(workflowId: string, eventName: string): boolean {
    const context = this.workflows.get(workflowId);
    if (context === undefined) return false;

    const workflow = this.definitions.get(workflowId);
    if (workflow === undefined) return false;

    return workflow.validateTransition(context.currentState, eventName, context) !== undefined;
  }

  /**
   * Return the names of transitions available from the current state.
   *
   * Transitions whose guards return `false` are excluded.
   */
  getAvailableEvents(workflowId: string): string[] {
    const context = this.workflows.get(workflowId);
    if (context === undefined) return [];

    const workflow = this.definitions.get(workflowId);
    if (workflow === undefined) return [];

    const transitions = workflow.getTransitions(context.currentState);
    return transitions
      .filter(t => t.guard === undefined || t.guard(context))
      .map(t => t.name);
  }

  /**
   * Suspend a running workflow.
   *
   * @throws {WorkflowError} If the workflow is not found.
   */
  async suspend(workflowId: string, reason: string = ''): Promise<void> {
    const context = this.workflows.get(workflowId);
    if (context === undefined) {
      throw new WorkflowError(`Unknown workflow: ${workflowId}`);
    }
    context.status = WorkflowStatus.Suspended;
    context.updatedAt = new Date();
    context.addHistoryEvent('suspended', { reason });
  }

  /**
   * Resume a suspended workflow.
   *
   * @throws {WorkflowError} If the workflow is not found or not suspended.
   */
  async resume(workflowId: string): Promise<void> {
    const context = this.workflows.get(workflowId);
    if (context === undefined) {
      throw new WorkflowError(`Unknown workflow: ${workflowId}`);
    }
    if (context.status !== WorkflowStatus.Suspended) {
      throw new WorkflowError(`Workflow ${workflowId} is not suspended`);
    }
    context.status = WorkflowStatus.Running;
    context.updatedAt = new Date();
    context.addHistoryEvent('resumed');
  }

  /**
   * Cancel a running workflow.
   *
   * @throws {WorkflowError} If the workflow is not found.
   */
  async cancel(workflowId: string, reason: string = ''): Promise<void> {
    const context = this.workflows.get(workflowId);
    if (context === undefined) {
      throw new WorkflowError(`Unknown workflow: ${workflowId}`);
    }
    context.status = WorkflowStatus.Cancelled;
    context.updatedAt = new Date();
    context.addHistoryEvent('cancelled', { reason });
  }

  /**
   * Get the current status of a workflow.
   *
   * @returns A {@link WorkflowResult}, or `null` if not found.
   */
  async getStatus(workflowId: string): Promise<WorkflowResult | null> {
    const context = this.workflows.get(workflowId);
    if (context === undefined) return null;

    const workflow = this.definitions.get(workflowId);
    const isFinal = workflow ? workflow.isFinalState(context.currentState) : false;

    let status = context.status;
    if (status === WorkflowStatus.Running && isFinal) {
      status = WorkflowStatus.Completed;
    }

    return {
      workflowId,
      status,
      currentState: context.currentState,
      output: Object.fromEntries(context.variables) as Record<string, unknown>,
      history: [...context.history],
      startedAt: context.startedAt,
      updatedAt: context.updatedAt,
    };
  }

  /**
   * Get the state change history for a workflow.
   *
   * @returns An array of history entries, or `null` if not found.
   */
  getHistory(workflowId: string): StateHistoryEntry[] | null {
    return this.histories.get(workflowId) ?? null;
  }

  private _recordHistory(workflowId: string, entry: Omit<StateHistoryEntry, 'timestamp'>): void {
    const history = this.histories.get(workflowId);
    if (history) {
      history.push({ ...entry, timestamp: new Date() });
    }
  }
}

/**
 * Factory function to create a new state machine definition.
 *
 * @param name - State machine name.
 * @returns A new {@link StateMachineBuilder} instance.
 */
export function stateMachine<T = unknown>(name: string): StateMachineBuilder<T> {
  return new StateMachineBuilder<T>(name);
}
